pkg/storage/storage.go

```diff
@@ -0,0 +1,191 @@
+package storage
+
+import (
+	"database/sql"
+	"fmt"
+
+	"github.com/google/uuid"
+	_ "github.com/mattn/go-sqlite3"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/carfloresf/the-sun-god/config"
+	"github.com/carfloresf/the-sun-god/internal/model"
+)
+
+const (
+	insertPostQuery = `INSERT
+						INTO post(id,title,author,subreddit,link,content,score,promoted,nsfw) 
+						VALUES    ($1, $2, $3, $4, $5, $6, $7, $8, $9)`
+	updatePostScoreQuery = `UPDATE post SET score = $1 WHERE id = $2`
+	countTotalPostsQuery = `SELECT COUNT(*) FROM post WHERE promoted = false`
+	queryPosts           = `SELECT id,title,author,subreddit,link,content,score,promoted,nsfw,created_at 
+							FROM post WHERE promoted = $1 ORDER BY score DESC, created_at DESC LIMIT $2 OFFSET $3;`
+	queryPromotedPosts = `SELECT id,title,author,subreddit,link,content,score,promoted,nsfw,created_at 
+							FROM post WHERE promoted = true ORDER BY RANDOM() LIMIT $1;`
+)
+
+var allStatements = []string{insertPostQuery, updatePostScoreQuery, countTotalPostsQuery, queryPosts, queryPromotedPosts}
+
+type DB struct {
+	Conn               *sql.DB
+	preparedStatements map[string]*sql.Stmt
+}
+
+func NewStorage(config *config.DB) (*DB, error) {
+	sqliteDatabase, err := sql.Open("sqlite3", config.DBFile)
+	if err != nil {
+		log.Fatal("error opening Conn connection: %w", err)
+		return nil, err
+	}
+
+	db := DB{
+		Conn: sqliteDatabase,
+	}
+
+	return &db, nil
+}
+
+func (d *DB) PrepareAllStatements() error {
+	d.preparedStatements = make(map[string]*sql.Stmt)
+
+	err := d.prepareStatements(allStatements)
+	if err != nil {
+		log.Errorln("error preparing statements:", err)
+		return err
+	}
+
+	return nil
+}
+
+func (d *DB) prepareStatements(queries []string) error {
+	for _, query := range queries {
+		stmt, err := d.Conn.Prepare(query)
```

Comment:
I might be wrong on this one, but why would you prepare all statements in advance? When you are using sql package with prepared statements, yes, you would have to prepare them every time, but the price of that might be okay compared to preparing them in advance. If you prepare statements in advance, they belong to a single DB connection, which means that multiple calls on the same prepared staements need to wait for that connection to be released. If you do not have benchmark showing this would increase your performance in important cases, I would rather prepare statements on the fly and leverage the ability to use any connection I can

Since this is sqlite, it makes sense, but you would be asked certainly about scalability and switching to another db for example. Then, the conn wait will be increased due to network latency also.

Feel free to ignore every comment if you feel like it :relaxed: ! I just want to provide different perspective

Reply:
Yeah as you mentioned for this case (sqlite) maybe is not that important, and they way I implemented it can be confusing, and a benchmark would be great in here (I could try to adding one). I'm not totally sure if the call waits for the connection I thought it was assigned another connection and "re-prepared" the statement (will need to check this in detail). 

In the past I have had good results using prepared statements (verified with load tests), but it was a different case than this one.

I thought that it was cool to do the "initial" development with sqlite, and in case it was required to switch to something else like postgres, so yeah it would be a good suggestion to improve if they ask about it.

So to wrap-up, maybe the best approach would be to test with a benchmark to be sure that it helps and have numbers to prove it, and remove prepared statements if they don't really help.